Reorganize translation references names

dragunoff opened the pull request:
The significant part here is the reorganization of the translation IDs in `en.ftl`. They are more descriptive to provide better context. Generally the template is _thing-id_ where "thing" is usually the UI element (label, button, notification, dialog, dropdown, checkbox, etc.). The aim of this PR is primarily to change the names. Later PRs may seek to consolidate some redundant strings and to better organize the translation file.

See this comment further down for more detail on the rationale behind the renaming and prefixing of keys: https://github.com/OpenRA/OpenRA/pull/20207#issuecomment-1239542263

~~Depends on #20048~~
~~Depends on #20400~~

PunkPun commented:
Are you planning to rebase or should we close this?

dragunoff commented:
Rebased.

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -117,469 +119,468 @@ overwrite-map-outside-edit-prompt = "The map has been edited from outside the ed
+label-paused = Paused
+label-max-speed = Max Speed
+label-percentage-speed = { $percentage }% Speed
+label-percentage-complete = { $percentage }% complete
```
These both feel undescriptive to me. Perhaps`label-replay-speed` and `label-replay-completness`?

PunkPun replied:
Though probably it's better to just remove the word `percantage`

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -117,469 +119,468 @@ overwrite-map-outside-edit-prompt = "The map has been edited from outside the ed
+label-teams-count = { $count } Teams
+label-humans-vs-bots = Humans vs Bots
+label-free-for-all = Free for all
+label-configure-teams = Configure Teams
```
these are buttons

dragunoff replied:
They appear as options in a dropdown but act like buttons... That's why I opted for the more generic "label" prefix here. But you have a point - essentially they are buttons, just dressed in a weird way.

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -117,469 +119,468 @@ overwrite-map-outside-edit-prompt = "The map has been edited from outside the ed
+label-slot = Slot
+label-open = Open
+label-closed = Closed
+label-bots = Bots
```
These are both buttons and labels. Should we care about that?

dragunoff replied:
I'd say we stick with "label" here.

PunkPun commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -82,73 +82,73 @@ public sealed class Server
 
 		[TranslationReference]
-		static readonly string TempBanned = "temp-banned";
+		static readonly string TempBanned = "notification-temp-banned";
```
perhaps 
```suggestion
		static readonly string TempBanned = "notification-temporarily-banned";
```

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -1,110 +1,112 @@
 
 ## ActorSelectorLogic
-type = Type
+tooltip-actor-type = Type: {$actorType}
```
```suggestion
tooltip-actor-type = Type: { $actorType }
```

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/WorldTooltipLogic.cs:
```diff
@@ -19,7 +19,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 	{
 		[TranslationReference]
-		static readonly string UnrevealedTerrain = "unrevealed-terrain";
+		static readonly string UnrevealedTerrain = "tooltip-unrevealed-terrain";
```
could the translation be cashed? It's called every frame

dragunoff replied:
I'd like to keep this PR focused on translation IDs and the FTL file structure. Let's address these kind of issues in separate PRs. Putting scope aside those are also more urgent and this PR might take a while to complete.

PunkPun replied:
Adressed in #20269

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/SpawnSelectorTooltipLogic.cs:
```diff
@@ -18,13 +18,13 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 		[TranslationReference("team")]
-		static readonly string TeamNumber = "team-number";
+		static readonly string TeamNumber = "label-team-name";
```
we query for team 0, even though we don't use it, since the label is visible only when `playerTeam > 0`. It seems incorrect

PunkPun replied:
Adressed in #20269

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngamePowerBarLogic.cs:
```diff
@@ -18,7 +18,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 	{
 		[TranslationReference]
-		static readonly string PowerUsage = "power-usage";
+		static readonly string PowerUsage = "tooltip-power-usage";
```
We query this every frame

PunkPun replied:
Adressed in #20269

PunkPun reviewed (commented):
I glossed though all the changes and looked for regressions in game. I noted down some things for this pr and a few less related optimisations

Mailaender commented:
Neither https://projectfluent.org/fluent/guide/ nor https://github.com/projectfluent/fluent/wiki/Good-Practices-for-Developers use any kind of redundant prefixes. I would recommend against adding those as well.

dragunoff commented:
> Neither https://projectfluent.org/fluent/guide/ nor https://github.com/projectfluent/fluent/wiki/Good-Practices-for-Developers use any kind of redundant prefixes. I would recommend against adding those as well.
>

These are contrived documentation examples and I think that approach doesn't scale well for a large project. I think we should have a look at how translation is handled in the wild. If we look at the [l10n files for Firefox](https://searchfox.org/mozilla-central/search?q=&path=.ftl&case=false&regexp=false) for example we can see plenty of redundancy in translation IDs (they use suffixes but that's besides the point). And the point is that this redundancy is there to provide context. I'll try to illustrate with an nice example from OpenRA's codebase:
```fluent
close = Close
```
Is that a verb or another part of speech (e.g. _"Close a dialog"_ or _"Close to something"_)? It lacks context so we can't really tell what is the meaning thus we can't provide adequate translation. In our case the meaning we're conveying is _"Camera zoom level that shows a close view of the battlefield"_. So I'd argue that adding context to the translation ID reduces ambiguity significantly and avoids conflicts (e.g. if we want to use "Close" as a button label). The translation ID can provide even more context by stating the type of UI element that this message is used in. So a better way to write this is:
```fluent
label-camera-close = Close
```

I can see how this quickly becomes quite verbose and repetitive so an even better way to write this may be with the help of [attributes](https://projectfluent.org/fluent/guide/attributes.html) (I don't know if they are supported in our implementation) and optionally [a comment bound to the translation](https://projectfluent.org/fluent/guide/comments.html):
```fluent
# A dropdown to choose the zoom level for the battlefield camera
dropdown-camera =
    .option-close = Close
    .option-medium = Medium
    .option-far = Far
    .option-furthest = Furthest
```

This approach works nicely for other UI elements that have a lot of related messages, such as dialogs:
```fluent
dialog-settings-restart =
    .title = Restart Now?
    .prompt = Some changes will not be applied until the game is restarted. Restart now?
    .accept = Restart Now
    .cancel = Restart Later
```

And to avoid some of the redundancy (for example with common button label names) we can pass a reference to another message:
```fluent
dialog-cant-play-video =
    .title = Unable to play video
    .prompt = Something went wrong during video playback.
    .accept = { button-back }

button-back = Back
```

I'd be fine with cutting down on some of the prefixes for messages that are not so context sensitive and conflicts are unlikely. But I don't see harm in keeping them and following a convention with regards to translation IDs (whether we use prefixes or suffixes doesn't matter much).

As the translation files are gonna grow larger we should think how to better organize them. Perhaps group translation by type of UI element instead of location in the source, maybe split the translation files and think of how we can generate them automatically from the source code :thinking:

dragunoff commented:
Update: rebased and addressed https://github.com/OpenRA/OpenRA/pull/20207#discussion_r962637302 and https://github.com/OpenRA/OpenRA/pull/20207#discussion_r962613625

PunkPun commented:
Needs a rebase

dragunoff commented:
Update: rebased and implemented attributes (see https://github.com/OpenRA/OpenRA/pull/20207#issuecomment-1239542263). I've expanded the translation linting to account for attributes and included a testcase commit that triggers all errors and warnings.

Mailaender commented:
I have split the attribute support and linting refactor into https://github.com/OpenRA/OpenRA/pull/20400 as I wanted to move forward. Hope you don't mind.

dragunoff commented:
> I have split the attribute support and linting refactor into #20400 as I wanted to move forward. Hope you don't mind.

Not at all :ok_hand: Thanks for picking this up. I'll rebase this on top of your PR.

dragunoff commented:
Update: rebased on #20400

dragunoff committed:
Refactor translation files

- Add prefixes to all message keys to provide context
- Use messages with attributes for some UI elements (dropdowns, dialogs, checkboxes, menus)
- Rename some class fields for consistency with translation keys

PunkPun commented:
Dependancy merged

dragunoff commented:
Update: rebased and updated some translation keys.

I decided to split the "dropdown" from its "options" because the options can be used in other contexts. So we have for example:

```fluent
dropdown-game-speed =
    .label = Game Speed
    .description = The rate at which time passes

options-game-speed =
    .slowest = Slowest
    .slower = Slower
    .normal = Normal
    .fast = Fast
    .faster = Faster
    .fastest = Fastest
```

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -1,730 +1,766 @@
-overwrite-map-outside-edit-prompt = "The map has been edited from outside the editor.
+dialog-overwrite-map-outside-edit =
+    .title = Warning
+    .prompt = "The map has been edited from outside the editor.
```
```suggestion
    .prompt = The map has been edited from outside the editor.
```

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -1,730 +1,766 @@
+    .prompt = There are too many players defined ({ $players }/{ $max }).
+    .confirm = Back
+
+
```
Double whitespace

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -1,730 +1,766 @@
+## LobbyLogic, InGameChatLogic
+button-general-chat = All
 
 ## LobbyLogic, IngameChatLogic
```
repeated twice

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -1,730 +1,766 @@
+    .options =
+        { $minutes ->
+            [one] { $minutes } minute
+        *[other] { $minutes } minutes
```
add more spaces?

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -23,40 +23,40 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 		[TranslationReference]
-		const string All = "all";
+		const string GeneralChat = "button-general-chat";
```
```suggestion
		const string AllChat = "button-all-chat";
```

Use gamer lingo?

dragunoff replied:
It's called "general chat" in the hotkeys so I went for consistency.
![image](https://user-images.githubusercontent.com/1355810/207067536-05f4e6fc-d3c9-420a-8731-db3e23ca539f.png)

PunkPun replied:
oh, perhaps we should change that

pchote replied:
IMO "Open General Chat" makes a lot more sense than "Open All Chat" (which could be interpreted as opening more than one chat dialog at the same time).

dragunoff replied:
I also think "general chat" is better and in any case this PR is not the place to change it.

dragunoff commented:
Update: addressed formatting nits

PunkPun commented:
other than that LGTM

pchote reviewed (approved):
LGTM, but a question before merging: I thought we wanted this on prep to avoid an unnecessary major API break between releases? Or has that changed?

PunkPun commented:
We do not want this for prep, unfortunately this didn't make the cut. This PR is based on #20400 which is also not on prep. 

The current translations on prep only work with UI and are still hardcoded to english, so only mods with custom UI might feel a change between releases, that is if they first decide to use the API. I think it's unlikely for the API break to have massive consequences so I'm fine with the state of linguini on prep

pchote commented:
The issue isn't the API changes to the UI, but to the translators. It is inevitable that, even if we warn against it, there are going to be people that put a lot of effort into creating translations only to find that their work is wasted/incompatible with the next release.

PunkPun commented:
That as well. Hopefully the lack of things to translate discourages people from doing it

PunkPun commented:
And limits the amount of wasted work / work to make it compatible again to a minimum

pchote commented:
I think thats a bit of a shame, but ok 👍

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/222afde629b5e42878e5b65bb4203deb2a3e88f3
